src/jit/gentree.cpp

```diff
@@ -9183,259 +9183,344 @@ GenTreePtr GenTree::GetChild(unsigned childNum)
     }
 }
 
-GenTreeUseEdgeIterator::GenTreeUseEdgeIterator() : m_node(nullptr), m_edge(nullptr), m_argList(nullptr), m_state(-1)
+GenTreeUseEdgeIterator::GenTreeUseEdgeIterator()
+    : m_advance(nullptr), m_node(nullptr), m_edge(nullptr), m_argList(nullptr), m_state(-1)
 {
 }
 
 GenTreeUseEdgeIterator::GenTreeUseEdgeIterator(GenTree* node)
-    : m_node(node), m_edge(nullptr), m_argList(nullptr), m_state(0)
+    : m_advance(nullptr), m_node(node), m_edge(nullptr), m_argList(nullptr), m_state(0)
 {
     assert(m_node != nullptr);
 
-    // Advance to the first operand.
-    ++(*this);
-}
-
-//------------------------------------------------------------------------
-// GenTreeUseEdgeIterator::GetNextUseEdge:
-//    Gets the next operand of a node with a fixed number of operands.
-//    This covers all nodes besides GT_CALL, GT_PHI, and GT_SIMD. For the
-//    node types handled by this method, the `m_state` field indicates the
-//    index of the next operand to produce.
-//
-// Returns:
-//    The node's next operand or nullptr if all operands have been
-//    produced.
-//
-GenTree** GenTreeUseEdgeIterator::GetNextUseEdge() const
-{
-    switch (m_node->OperGet())
+    if (m_node->OperIsConst() || m_node->OperIsLeaf())
     {
-        case GT_CMPXCHG:
-            switch (m_state)
+        m_state = -1;
+    }
+    else if (m_node->OperIsUnary())
+    {
+        if (m_node->OperIs(GT_PHI))
+        {
+            m_argList = m_node->AsUnOp()->gtOp1;
+            m_advance = &GenTreeUseEdgeIterator::AdvanceList;
+        }
+        else
+        {
+            GenTree** edge = &m_node->AsUnOp()->gtOp1;
+            if (*edge == nullptr)
             {
-                case 0:
-                    return &m_node->AsCmpXchg()->gtOpLocation;
-                case 1:
-                    return &m_node->AsCmpXchg()->gtOpValue;
-                case 2:
-                    return &m_node->AsCmpXchg()->gtOpComparand;
-                default:
-                    return nullptr;
+                m_state = -1;
             }
-        case GT_ARR_BOUNDS_CHECK:
-#ifdef FEATURE_SIMD
-        case GT_SIMD_CHK:
-#endif // FEATURE_SIMD
-            switch (m_state)
+            else
             {
-                case 0:
-                    return &m_node->AsBoundsChk()->gtIndex;
-                case 1:
-                    return &m_node->AsBoundsChk()->gtArrLen;
-                default:
-                    return nullptr;
+                m_edge    = edge;
+                m_advance = &GenTreeUseEdgeIterator::AdvanceUnary;
             }
+            return;
+        }
```

Comment:
This case terminates early to skip the call to `++(*this)` at the end of the constructor, as unary nodes are already in the proper entry state. We could (and should) do the same for leaf nodes.